File: Python/programmers/etc/level3/공_이동_시뮬레이션.py
def solution(n, m, x, y, queries):
    sr, er, sc, ec = x, x, y, y
    queries = queries[::-1]
    for d, dx in queries:
        if d == 0:
            if sc > 0:
                sc += dx
            ec = min(m-1, ec+dx);
        elif d == 1:
            if ec < m-1:
                ec -= dx
            sc = max(0, sc-dx)
        elif d == 2:
            if sr > 0:
                sr += dx
            er = min(n-1, er+dx)
        else:
            if er < n-1:
                er -= dx
            sr = max(0, sr-dx)
        
        if sr>=n or er<0 or sc>=m or ec<0:
            return 0

    return (er-sr+1)*(ec-sc+1)

File: Java/programmers/etc/level3/공이동시뮬레이션.java
package CodeTest.Java.programmers.etc.level3;

public class 공이동시뮬레이션 {
    public long solution(int n, int m, int x, int y, int[][] queries) {
        long sr=x, sc=y, er=x, ec=y;

		for(int i=queries.length-1 ; i>=0 ; i--) {
			int command=queries[i][0], dx=queries[i][1];
			
			if(command==0) {
				if(sc>0) sc += dx;
				ec = Math.min(m-1, ec+dx);
			} 
            else if(command==1) {
				if(ec<m-1) ec -= dx;
				sc = Math.max(0, sc-dx);
			} 
            else if(command==2) {
				if(sr>0) sr += dx;
				er = Math.min(n-1, er+dx);
			} 
            else {
				if(er<n-1) er -= dx;
				sr = Math.max(0, sr-dx);
			}
			if(sr>=n || er<0 || sc>=m || ec<0) return 0;	
		}
		return (er-sr+1)*(ec-sc+1);
    }
}
